scheduler/src/cook/plugins/submission.clj

```diff
@@ -112,15 +113,129 @@
         do-one-job (fn do-one-job [job-map]
                      (let [now (t/now)
                            status (if (t/before? now deadline)
-                                    (check-job-submission plugin-object job-map)
+                                    (check-job-submission plugin-object job-map pool-name)
                                     ; Running out of time, do the default.
                                     (check-job-submission-default plugin-object))]
                        (or status default-accept)))
         results (map do-one-job jobs)
         errors (filter #(= :rejected (:status %)) results)
         error-count (count errors)
         ; Collect a few errors to show in the response. (not every error)
-        error-samples (apply list (take 3 errors))]
+        max-errors-to-show 3
+        error-samples (->> errors
+                           (map :message)
+                           (take max-errors-to-show)
+                           (apply list))]
     (if (zero? error-count)
       {:status :accepted}
-      {:status :rejected :message (str "Total of " error-count " errors. First 3 are: " error-samples)})))
+      {:status :rejected
+       :message
+       (str
+         (if (<= error-count max-errors-to-show)
+           ""
+           (str "Total of " error-count " errors; first "
+                (count error-samples) " are:\n"))
+         (str/join "\n" error-samples))})))
+
+(defrecord JopShapeValidationPlugin
+  []
+  JobSubmissionValidator
+
+  (check-job-submission-default [_]
+    {:status :accepted})
+
+  (check-job-submission [_ {:keys [constraints cpus disk gpus mem]} pool-name]
+    (if (and gpus (pos? gpus))
```

Comment:
so why are gpu jobs separate?

Reply:
Because the GPU model and number of GPUs requested become inputs to the validation in addition to node-type, node-family, and cpu-architecture.

Reply:
yeah, that happens during node matching?
why is the node type matching ignored if you ask for GPUs?